### PR TITLE
pack `TypeFunction` flags in a bit set to reduce memory use

### PR DESCRIPTION
**Pull request by ghost**
This saves 3MB on a project of mine that's ~10K SLOCS, or 2MB on dparse (but dparse does not contain that much functions).
I see no significant differences on the user time (my method to test is `/usr/bin/time -v ...` ).

As said in the forum there's no easy gain anymore. My previous PRs that allowed to reduce by a couple of MB were small. This time the diff is big, but mostly because the changes have to be copied and paste in AstBase and the cpp header.

### CODE REVIEW

File: src/dmd/astbase.d [hunk 1]
```diff
@@ -3933,21 +3933,33 @@ struct ASTBase
...
-        bool isreturn;              // true: 'this' is returned by ref
-        bool isscope;               // true: 'this' is scope
-        bool islive;                // true: function is @live
+        // .next is the return type
+
+        ParameterList parameterList;   // function parameters
+
+        private enum FunctionFlag
+        {
+            isnothrow       = 0x0001, // nothrow
```
Comment (MoonlightSentinel) [inline src/dmd/astbase.d:3942]: Might as well add a `/` to turn those into DDoc comments.

File: src/dmd/astbase.d [hunk 2]
```diff
@@ -3995,13 +4007,165 @@ struct ASTBase
...
+        }
+        /// ditto
+        void isreturn(bool v)
+        {
+            if (v) funcFlags |= FunctionFlag.isreturn;
+            else funcFlags &= ~FunctionFlag.isreturn;
+        }
+
+        /// set or get if the function has the `scope` attribute
+        bool isthisscope() const pure
```
Comment (MoonlightSentinel) [inline src/dmd/astbase.d:4081]: Why add `this`?

File: src/dmd/mtype.d [hunk 3]
```diff
@@ -4517,11 +4525,12 @@ extern (C++) final class TypeFunction : TypeNext
...
             tf.isref = t.isref;
             tf.isreturn = t.isreturn;
-            tf.isscope = t.isscope;
+            tf.isthisscope = t.isthisscope;
             tf.isreturninferred = t.isreturninferred;
             tf.isscopeinferred = t.isscopeinferred;
             tf.trust = t.trust;
-            tf.iswild = t.iswild;
+            tf.isInOutParam = t.isInOutParam;
+            tf.isInOutQual = t.isInOutQual;
```
Comment (MoonlightSentinel) [inline src/dmd/mtype.d:4533]: Unrelated change?

File: src/dmd/mtype.d [hunk 4]
```diff
@@ -4581,10 +4590,11 @@ extern (C++) final class TypeFunction : TypeNext
         t.isproperty = isproperty;
         t.isref = isref;
         t.isreturn = isreturn;
-        t.isscope = isscope;
+        t.isthisscope = isthisscope;
         t.isreturninferred = isreturninferred;
         t.isscopeinferred = isscopeinferred;
-        t.iswild = 0;
+        t.isInOutParam = false;
+        t.isInOutQual = false;
```
Comment (MoonlightSentinel) [inline src/dmd/mtype.d:4597]: Ditto?

File: src/dmd/mtype.d [hunk 5]
```diff
@@ -4169,22 +4169,29 @@ extern (C++) final class TypeFunction : TypeNext
...
+        isreturn        = 0x0010, // 'this' is returned by ref
+        isscope         = 0x0020, // 'this' is scope
+        isreturninferred= 0x0040, // 'this' is return from inference
+        isscopeinferred = 0x0080, // 'this' is scope from inference
+        islive          = 0x0100, // is @live
+        incomplete      = 0x0200, // return type or default arguments removed
+        inoutParam      = 0x0400, // inout on the parameters
+        inoutQual       = 0x0800, // inout on the qualifier
+    }
+    static assert (FunctionFlag.max <= 0x8000, "flags are stored in a ushort");
```
Comment (MoonlightSentinel) [inline src/dmd/mtype.d:4187]: Could also make this more explicit:

```suggestion
    static assert (FunctionFlag.max <= ushort.max, "flags are stored in a ushort");
```

File: src/dmd/mtype.d [hunk 6]
```diff
@@ -5010,6 +5020,155 @@ extern (C++) final class TypeFunction : TypeNext
         return false;
     }
 
+    /// set or get if the function has the `nothrow` attribute
+    bool isnothrow() const pure
```
Comment (MoonlightSentinel) [inline src/dmd/mtype.d:5024]: What about `@safe`, `nothrow` and `@nogc`?

**Review by MoonlightSentinel — CHANGES_REQUESTED** [review]

File: src/dmd/astbase.d (hunk 2 above)
Comment (NilsLankila) [reply src/dmd/astbase.d:4081]: Good question. It's because `Type.isscope()` already exists and has a complete different meaning. That was somewhat the big problem when writing this PR...

File: src/dmd/mtype.d (hunk 3 above)
Comment (NilsLankila) [reply src/dmd/mtype.d:4533]: previously the two values were packaed in the `iswild` byte. now they are part of the new flags.

File: src/dmd/mtype.d (hunk 5 above)
Comment (NilsLankila) [reply src/dmd/mtype.d:4187]: No, in this case you cannot have a valid enum member over 0x8000, remember, it's for a set so everything that between 0x8001 and 0xFFFF would cause problems.

Comment (MoonlightSentinel) [reply src/dmd/mtype.d:4187]: I know, but then you would also have to check that there are no values between 0x4001 and 0x7FFF, ... (additionally there's the precedent of `STC.safeGroup` and friends).
This assert effectively checks that 1 is equal to 1 ATM.

File: src/dmd/astbase.d (hunk 1 above)
Comment (NilsLankila) [reply src/dmd/astbase.d:3942]: it's private and it's just a copy of the original code comments. better DDOC is in the public accessors.

File: src/dmd/astbase.d (hunk 2 above)
Comment (MoonlightSentinel) [reply src/dmd/astbase.d:4081]: That's unfortunate. Not a big fan of `this` but can't think of a better alternative right now.

File: src/dmd/astbase.d (hunk 1 above)
Comment (MoonlightSentinel) [reply src/dmd/astbase.d:3942]: `private` doesn't really matter IMHO because DDoc comments are shown in IDE's anyway.

But i guess it  doesn't matter as much in this case.

File: src/dmd/astbase.d (hunk 2 above)
Comment (NilsLankila) [reply src/dmd/astbase.d]: yeah. The worst is that the Type.isscope is mostly a noop but for classes.

File: src/dmd/astbase.d [hunk 7]
```diff
@@ -3995,13 +4007,165 @@ struct ASTBase
...
+            t.isscopeinferred = isscopeinferred;
+            t.isInOutParam = isInOutParam;
+            t.isInOutQual = isInOutQual;
             t.trust = trust;
             t.fargs = fargs;
             return t;
         }
 
+        /// set or get if the function has the `nothrow` attribute
+        bool isnothrow() const pure nothrow @safe @nogc
```
Comment (n8sh) [inline src/dmd/astbase.d:4021]: Won't accessor functions eat up the saved space if you don't mark them `final`, or do `extern(C++)` objects not carry around function pointers the way `extern(D)` objects do?

Comment (NilsLankila) [reply src/dmd/astbase.d:4021]: they are final because `TypeFunction` is `final class`

File: src/dmd/clone.d [hunk 8]
```diff
@@ -315,7 +315,7 @@ FuncDeclaration buildOpAssign(StructDeclaration sd, Scope* sc)
         auto idswap = Identifier.generateId("__swap");
         auto swap = new VarDeclaration(loc, sd.type, idswap, new VoidInitializer(loc));
         swap.storage_class |= STC.nodtor | STC.temp | STC.ctfe;
-        if (tdtor.isscope)
+        if (tdtor.isthisscope)
```
Comment (n8sh) [inline src/dmd/clone.d:318]: Was the old name misleading?

Comment (MoonlightSentinel) [reply src/dmd/clone.d:318]: See https://github.com/dlang/dmd/pull/11461#discussion_r460561309

File: src/dmd/astbase.d [hunk 9]
```diff
@@ -3995,13 +4007,165 @@ struct ASTBase
...
+        {
+            return (funcFlags & FunctionFlag.isnothrow) != 0;
+        }
+        /// ditto
+        void isnothrow(bool v) nothrow @safe @nogc
```
Comment (MoonlightSentinel) [inline src/dmd/astbase.d:4026]: Still missing `pure`

Comment (NilsLankila) [reply src/dmd/astbase.d]: done

**Comment by UplinkCoder** [comment]
This is a fine example of a PR, which is highly unlikely to introduce regressions.
Good work @NilsLankila

**Comment by NilsLankila** [comment]
@UplinkCoder, in case your remark was irony, function attributes are well tested. I've been able to verify that for example when I changed `.syntaxcopy()` so that it copied the whole set "container" instead of individually set the set members. When I did that it broke druntime & phobos. I'm quite confident on the correctness of the refact however I admit it can appear scary.

**Comment by UplinkCoder** [comment]
@NilsLankila It was not!
This is how it should be done.

**Comment by NilsLankila** [comment]
thanks for the clarification @UplinkCoder.

**Review by UplinkCoder — APPROVED** [review]
Nice one.

**Comment by Geod24** [comment]
Looking at this PR, it is fairly obvious we need a `Flag` data structure...

File: src/tests/cxxfrontend.c [hunk 10]
```diff
@@ -351,14 +351,14 @@ void test_types()
...
-    assert(tfunction->isreturninferred);
-    assert(!tfunction->isscopeinferred);
+    assert(tfunction->isnothrow());
+    assert(!tfunction->isnogc());
+    assert(tfunction->isproperty());
+    assert(!tfunction->isref());
+    assert(tfunction->isreturn());
+    assert(!tfunction->isthisscope());
+    assert(tfunction->isreturninferred());
+    assert(!tfunction->isscopeinferred());
```
Comment (ibuclaw) [inline src/tests/cxxfrontend.c:361]: Could you add tests for the setters too?  One ought to be enough, e.g:
```
tfunction->isref(true);
assert(tfunction->isref());
```

Comment (NilsLankila) [reply src/tests/cxxfrontend.c:363]: done

**Review by Geod24 — APPROVED** [review]

File: src/dmd/astbase.d [hunk 11]
```diff
@@ -3933,21 +3933,33 @@ struct ASTBase
...
-        ParameterList parameterList;  // function parameters
-
-        bool isnothrow;             // true: nothrow
-        bool isnogc;                // true: is @nogc
-        bool isproperty;            // can be called without parentheses
-        bool isref;                 // true: returns a reference
-        bool isreturn;              // true: 'this' is returned by ref
-        bool isscope;               // true: 'this' is scope
-        bool islive;                // true: function is @live
+        // .next is the return type
```
Comment (MoonlightSentinel) [inline src/dmd/astbase.d:3936]: This seems out of place?

File: src/dmd/astbase.d [hunk 12]
```diff
@@ -3995,13 +4007,165 @@ struct ASTBase
...
+        {
+            return (funcFlags & FunctionFlag.inoutQual) != 0;
+        }
+        /// ditto
+        void isInOutQual(bool v) pure nothrow @safe @nogc
+        {
+            if (v) funcFlags |= FunctionFlag.inoutQual;
+            else funcFlags &= ~FunctionFlag.inoutQual;
+        }
+        /// Returns: wether the function is isInOutQual or isInOutParam
```
Comment (MoonlightSentinel) [inline src/dmd/astbase.d:4163]: Typo:
```suggestion
        /// Returns: whether the function is isInOutQual or isInOutParam
```

File: src/dmd/mtype.d [hunk 13]
```diff
@@ -5010,6 +5020,155 @@ extern (C++) final class TypeFunction : TypeNext
...
+    {
+        return (funcFlags & FunctionFlag.inoutQual) != 0;
+    }
+    /// ditto
+    void isInOutQual(bool v) pure nothrow @safe @nogc
+    {
+        if (v) funcFlags |= FunctionFlag.inoutQual;
+        else funcFlags &= ~FunctionFlag.inoutQual;
+    }
+    /// Returns: wether the function is isInOutQual or isInOutParam
```
Comment (MoonlightSentinel) [inline src/dmd/mtype.d:5166]: ditto

File: src/dmd/astbase.d [hunk 14]
```diff
@@ -3995,13 +4007,165 @@ struct ASTBase
...
+        }
+        /// ditto
+        void isreturn(bool v) pure nothrow @safe @nogc
+        {
+            if (v) funcFlags |= FunctionFlag.isreturn;
+            else funcFlags &= ~FunctionFlag.isreturn;
+        }
+
+        /// set or get if the function has the `scope` attribute
+        bool isthisscope() const pure nothrow @safe @nogc
```
Comment (MoonlightSentinel) [inline src/dmd/astbase.d:4081]: Might as well use camelCase since this introduces a new symbol which breaks the naming convention.

**Review by MoonlightSentinel — APPROVED** [review]
Otherwise LGTM

File: src/dmd/astbase.d (hunk 14 above)
Comment (NilsLankila) [reply src/dmd/astbase.d:4081]: I'll go for `isScopeQual`, as there's l`IsInOutQuall` already.

File: src/dmd/astbase.d (hunk 11 above)
Comment (NilsLankila) [reply src/dmd/astbase.d:3936]: no it's just a hint coming from the real AST that explain with TypeClass decrivates from TypeNext
